bokehjs/src/lib/models/glyphs/varea_step.ts

```diff
@@ -92,28 +92,28 @@ export class VAreaStepView extends AreaView {
 
   protected override _hit_point(geometry: PointGeometry): Selection {
     const {sx, sy1, sy2} = this
+    const step_mode = this.model.step_mode
 
-    for (let i = 0; i < this.data_size - 1; i++) {
+    for (let i = 0; i < this.data_size; i++) {
       let px: number[]
-      let py: number[]
 
-      switch (this.model.step_mode) {
-        case "before":
-          px = [sx[i], sx[i+1], sx[i+1], sx[i]]
-          py = [sy1[i+1], sy1[i+1], sy2[i+1], sy2[i+1]]
-          break
-        case "after":
-          px = [sx[i], sx[i+1], sx[i+1], sx[i]]
-          py = [sy1[i], sy1[i], sy2[i], sy2[i]]
-          break
-        case "center":
-        {
-          const mid_x = (sx[i] + sx[i+1]) / 2
-          px = [sx[i], mid_x, mid_x, sx[i+1], sx[i+1], mid_x, mid_x, sx[i]]
-          py = [sy1[i], sy1[i], sy1[i+1], sy1[i+1], sy2[i+1], sy2[i+1], sy2[i], sy2[i]]
-          break
+      if (step_mode == "before" && i != 0)
+        px = [sx[i-1], sx[i], sx[i], sx[i-1]]
+      else if (step_mode == "after" && i != this.data_size - 1)
+        px = [sx[i], sx[i+1], sx[i+1], sx[i]]
+      else if (step_mode == "center") {
+        const mid_prev_x = (sx[i-1] + sx[i]) / 2
+        const mid_next_x = (sx[i] + sx[i + 1]) / 2
+        if (i == 0) {
+          px = [sx[i], mid_next_x, mid_next_x, sx[i]]
+        } else if (i == this.data_size - 1) {
+          px = [mid_prev_x, sx[i], sx[i], mid_prev_x]
+        } else {
+          px = [mid_prev_x, mid_next_x, mid_next_x, mid_prev_x]
```

Comment:
`step_mode` is an enumeration, so we should really prefer a `switch` statement over and `if-elseif` chain, so that the compiler can validate that all cases are always covered. @solstag As long as this code is getting updated, it would be good to clean it up more thoroughly. Are you available to work on it a little more? 

As I see it the main complexity is that all three cases have different loop bounds, so the best solution would seem to be to not do that, and use helper functions in each case:
```typescript
    switch (this.model.step_mode) {
      case "before":
        px = this._hit_point_before()
      case "after":
        px = this._hit_point_after()
      case "center":
        px = this._hit_point_center()
```
Each helper function can have an appropriate loop inside, with the correct start/end so that no ` && i != this.data_size - 1` type checks are needs. As a bonus there can just be an immediate return inside the loop, so not mutable `let` variables or `break` would be needed either.

Reply:
Actually maybe even better to define a `self._line_selection_for(i)` helper to make selection, then just have the mode helpers return a selection directly. Something like:
```typescript
protected _hit_point_before(geometry: PointGeometry): Selection {
  for (let i = 1; i < this.data_size; i++) {
    const px = [sx[i-1], sx[i], sx[i], sx[i-1]]
    const py = [sy1[i], sy1[i], sy2[i], sy2[i]] 

    if (hittest.point_in_poly(geometry.sx, geometry.sy, px, py)) {
      return self._line_selection_for(i)

    return new Selection()
```

Then the switch can just become:

```typescript
    switch (this.model.step_mode) {
      case "before":
        return this._hit_point_before(geometry)
      case "after":
        return this._hit_point_after(geometry)
      case "center":
        return this._hit_point_center(geometr)
```
Let me know what you think.